File: Part 6 - Reinforcement Learning/Section 32 - Upper Confidence Bound (UCB)/upper_confidence_boundPRACTICE.R
dataset = read.csv('Ads_CTR_Optimisation.csv')
N= 10000
d=10
no_of_selections = integer(d)
sums_of_rewards= integer(d)
ads_selected = integer()
for (n in 1:N){
  if (n<=10){
    ads_selected = append(ads_selected, n)
    no_of_selections[n] = no_of_selections[n] + 1
    sums_of_rewards[n] = sums_of_rewards[n] + dataset[n,n]
  }

  else{
    upper_bound = integer()
    for (i in 1:d){
      upper_bound = append(upper_bound, sums_of_rewards[i]/no_of_selections[i]+ sqrt(1.5*log(n)/no_of_selections[i]))
    }
    next_index = match(max(upper_bound),upper_bound)
    no_of_selections[next_index] = no_of_selections[next_index]+ 1
    sums_of_rewards[next_index]= sums_of_rewards[next_index]+ dataset[n,next_index]
    ads_selected = append(ads_selected, next_index)
  }
}
total_rewards = sum(sums_of_rewards)

hist(ads_selected)

File: Traditional Machine Learning Models/Reinforcement Learning/Upper Confidence Bound (UCB)/upper_confidence_boundPRACTICE.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import math

dataset = pd.read_csv('Ads_CTR_Optimisation.csv')
N = 10000
d = 10
ad_selected = []
rewards = [0]*d
no_of_selections = [0]*d
total_reward =0
for i in range(0,N):
       if i<=9:
              rewards[i] += dataset.values[i,i]
              no_of_selections[i]+=1
              ad_selected.append(i+1)
       else:
              upper_bound =[]
              for j in range(0,d):
                     upper_bound.append(rewards[j]/no_of_selections[j]+ math.sqrt(1.5*math.log(i+1)/no_of_selections[j]))
              next_index = upper_bound.index(max(upper_bound))
              no_of_selections[next_index]+=1
              rewards[next_index]+= dataset.values[i,next_index]
              ad_selected.append(next_index+1)
total_reward = sum(rewards)
plt.hist(ad_selected)
plt.xlabel('Ad number')
plt.ylabel('No of selections')
plt.title('Histogram')
plt.show()

actual_max = sum(dataset.iloc[:,4].values)      
